File: libs/data/queue.go
package data

type (
	Queue struct {
		start, end *node
		length     int
	}

	node struct {
		value interface{}
		next  *node
	}
)

func NewQueue() *Queue {
	return &Queue{nil, nil, 0}
}

func (q *Queue) Enqueue(v interface{}) {
	n := &node{v, nil}
	if q.length == 0 {
		q.start = n
		q.end = n
	} else {
		q.end.next = n
		q.end = n
	}
	q.length++
}

func (q *Queue) Dequeue() interface{} {
	if q.length == 0 {
		return nil
	}
	n := q.start
	if q.length == 1 {
		q.start = nil
		q.end = nil
	} else {
		q.start = q.start.next
	}
	q.length--
	return n.value
}

func (q *Queue) Len() int {
	return q.length
}

func (q *Queue) Peek() interface{} {
	if q.length == 0 {
		return nil
	}
	return q.start.value
}

func (q *Queue) IsEmpty() bool {
	if q.length == 0 {
		return true
	}
	return false
}

File: leetcode/queue.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

const int MAX = 100000;

struct Node {
  Node* next;
  string name;

  Node()
}

int qu[MAX];
int tail = 0, head = 0;

void init() {
  head = tail = 0;
}

bool isEmpty() {
  return (head == tail);
}

bool isFull() {
  return (head == (tail + 1) % MAX);
}

void enqueue(int x) {
  if (isFull()) {
    cout << "error: queue is full" << endl;
    return;
  }

  qu[tail] = x;
  ++tail;
  if(tail == MAX) tail = 0;
}

int dequeue() {
  if(isEmpty()){
    cout << "error: queue is empty" << endl;
    return;
  }
  int res = qu[head];
  ++head;
  if(head==MAX) head = 0;
  return res;
}

int main() {
  init();

  enqueue(3);
  enqueue(5);
  enqueue(9);

  cout << qu << endl;
  cout << dequeue() << endl;

  cout << qu << endl;
  cout << dequeue() << endl;
  
  enqueue(33);
  cout << qu << endl;  
}
